### DISCUSSION TITLE
Use of guid as primary key and scafold crud with sergen

### DISCUSSION THREAD

**Discussion by elenaaralla** [serenity-is/Serenity#6730] [General]
Hi,

I created a table with primary key field of type Guid using FluentMigrator, i.e.:
```
this.Create.Table("Tenants")
    .WithColumn("Id").AsGuid().NotNullable().PrimaryKey()
    .WithColumn("Name").AsString(100).NotNullable()
    .WithColumn("Logo").AsString(100).Nullable();
```

I generated code with sergen and added field of type Guid to TenantForm.cs (with hidden decoration to avoid to show field on Create new form.

If I save new record, I get an error due to null value of EntityId; so the question is:

Is it possible define guid primary keys?
How have I to change generated code to make it works?

**Comment by dfaruque** [comment]
try `.WithDefault(SystemMethods.NewGuid)`
```
this.Create.Table("Tenants")
    .WithColumn("Id").AsGuid().PrimaryKey().WithDefault(SystemMethods.NewGuid)
    .WithColumn("Name").AsString(100).NotNullable()
    .WithColumn("Logo").AsString(100).Nullable();
```

**Reply by elenaaralla** [reply]
> .WithDefault(SystemMethods.NewGuid)

Hi,

it's work on migration, but when I try to insert a new tenant from application, the filed Id of tenant is null; I thing there is a problem with the cast of form values (that are strings) and type on tenantRow  class property id that is a Guid?

**Reply by dfaruque** [reply]
try assigning it in the save handler

```
    public class SampleTransactionSaveHandler : SaveRequestHandler<MyRow, MyRequest, MyResponse>, ISampleTransactionSaveHandler
    {
        public SampleTransactionSaveHandler(IRequestContext context)
             : base(context)
        {
        }

        protected override void BeforeSave()
        {
            base.BeforeSave();

            Row.Id ??= Guid.NewGuid();
        }
    }
```

**Reply by elenaaralla** [reply]
> try assigning it in the save handler
> 
> ```
>     public class SampleTransactionSaveHandler : SaveRequestHandler<MyRow, MyRequest, MyResponse>, ISampleTransactionSaveHandler
>     {
>         public SampleTransactionSaveHandler(IRequestContext context)
>              : base(context)
>         {
>         }
> 
>         protected override void BeforeSave()
>         {
>             base.BeforeSave();
> 
>             Row.Id ??= Guid.NewGuid();
>         }
>     }
> ```

Hi,

your suggestion applies only when update entity, not when create a new one.

I solved adding field id on TenatForm.cs 

```
public class TenantForm.cs
{
    [Hidden]
    public Guid? Id { get; set; }
    public string Name { get; set; }
    public string Logo { get; set; }
}
```

and adding this code to TenantDialog.ts:

```
...
protected form = new TenantForm(this.idPrefix);

protected updateInterface() {
    super.updateInterface();

    if (this.isNew()) {
        this.form.Id.value = Guid.create().toString();
    }
}
```

but I'm not sure it is the right mode to solve the error on inserting new tenant
thank you,
Elena
